overlord/devicestate/devicestate.go

```diff
@@ -433,7 +435,47 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
 		(ms.newModelSnap.SnapType == "kernel" || ms.newModelSnap.SnapType == "gadget")
 }
 
-func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSnapsForRemodel, deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
+type remodelOptions struct {
+	localSnapsRequired bool
+}
+
+func (ro *remodelOptions) UpdateWithDeviceContext(st *state.State,
+	localSi *snap.SideInfo, path, snapName, channel string,
+	userID int, snapStateFlags snapstate.Flags,
+	deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
+	logger.Debugf("snap %s track changed", snapName)
+	if ro.localSnapsRequired {
+		if localSi == nil {
+			return nil, fmt.Errorf("no snap file provided for %q (track changed)", snapName)
+		}
+		return snapstateUpdatePathWithDeviceContext(st, localSi, path, snapName,
+			&snapstate.RevisionOptions{Channel: channel},
+			userID, snapStateFlags, deviceCtx, fromChange)
+	}
+	return snapstateUpdateWithDeviceContext(st, snapName,
+		&snapstate.RevisionOptions{Channel: channel},
+		userID, snapStateFlags, deviceCtx, fromChange)
+}
+
+func (ro *remodelOptions) InstallWithDeviceContext(ctx context.Context, st *state.State,
+	localSi *snap.SideInfo, path, snapName, channel string, userID int,
```

Comment:
Couldn't some of these many parameters be a part of `remodelOptions`?

Reply:
See the other related comment.